### DISCUSSION TITLE
Wagmi v1 Interoperability with Ethers Dependent Libraries

### DISCUSSION THREAD

**Discussion by frankfka** [wevm/wagmi#2340] [Q&A] [answered]
Hello,

With the change to `viem` dependency, is it no longer to interoperate with ethers? For example, given a `WalletClient`, how could I use another npm dependency that requires an ethers `Provider` or `Signer`?

Additionally, is there any guidance for building custom connectors when the underlying libraries use ethers?

**Comment by jxom** [comment]
You can convert a `WalletClient` into a `Provider` or `Signer` quite trivially:

```ts
import { createWalletClient, custom } from 'viem'
import { mainnet } from 'viem/chains'
import { providers } from 'ethers'

const account = '0x0000000000000000000000000000000000000000'

const client = createWalletClient({ 
  account, 
  chain: mainnet,
  transport: custom(window.ethereum)
})

// ethers.js
const provider = new providers.Web3Provider(window.ethereum, client.chain)
const signer = provider.getSigner(client.account.address)
```

**Reply by frankfka** [reply]
Thanks for the reply! So seeing a few TS errors. I'm getting `publicClient` from `usePublicClient`:

```
import 'wagmi/window'; 

...

// Property 'ethereum' does not exist on type 'typeof import("node_modules/wagmi/dist/window")'
// Argument of type 'Chain' is not assignable to parameter of type 'Networkish | undefined'.   Property 'chainId' is missing in type 'Chain' but required in type 'Network'
new providers.Web3Provider(window.ethereum, publicClient.chain);
```

Furthermore, even with the `import 'wagmi/window'; ` - is there no explicit consideration for SSR?

**Reply by frankfka** [reply]
@jxom - wanted to follow up on this

**Reply by tmm** [reply]
Can you create a minimal reproduction?

**Reply by frankfka** [reply]
Playing around in Stackblitz now, it seems to pass TS, though I was wondering if we could do something like:

```ts
const provider = new providers.Web3Provider(
  publicClient,
  walletClient.chainId
);
```
Or, do we need that explicit import for `wagmi/window` and pass `window.ethereum` directly?

**Reply by frankfka** [reply]
On another note though, I can't find any examples to go from a `WalletClient` -> ethers `Wallet`, is there any guidance for that?

**Comment by frankfka** [comment]
Additionally, wanted to follow up on converting a `Signer` -> `WalletClient` - we have our own custom connector that exposes a `Signer`, but now I'll need to map that to a `WalletClient`

**Reply by zkSoju** [reply]
Looking for this as well as I need `Wallet` to use SnapshotJS
https://github.com/snapshot-labs/snapshot.js/blob/master/src/sign/index.ts#L62

https://docs.ethers.org/v6/api/wallet/#Wallet

**Comment by 0xjojoex** [comment]
sharing what we've done so far
```js
const { connector } = useAccount()

// get the connected provider instead of window.ethereum
connector.getProvider().then((provider) => {
  return new Web3Provider(provider).getSigner()
})
```

For snapshot.js, we can simply mocking the web3 if you don't need the whole signer
```ts
const web3 = {
  getSigner: () => {
    return {
      _signTypedData: (domain, types, message) =>
        signer.signTypedData({
          account,
          domain,
          types,
          message,
          primaryType: 'Vote',
        }),
    }
  },
}

const client = new snapshot.Client712(hub)

await client.vote(web3 as any, account, {
  space: PANCAKE_SPACE,
  choice: vote.value,
  reason: '',
  type: 'single-choice',
  proposal: proposalId,
  app: 'snapshot',
})
```

**Comment by artschekoff** [comment]
I have a lot of links to ethers, also because wagmi was based on it. Now huge radical changes, tons of replacements and I have to rewrite everything completely?

Is this appropriate behavior for reliable lib? Whats with ethers? Is this abandoned and impossible to base on?

**Comment by chybisov** [comment]
Example of how to convert `viem` -> `ethers` and `ethers` -> `viem`. 

```ts
import { providers } from 'ethers';
import { createWalletClient, custom } from 'viem';

// viem to ethers

const client0 = createWalletClient({
  transport: custom(window.ethereum),
});

// provide client0.chain.id or 'any' to accept flawless chain switch
const provider = new providers.Web3Provider(client0.transport, 'any');
const signer = provider.getSigner();

console.log(await signer.getAddress(), await signer.getChainId());

// ethers to viem

const client1 = createWalletClient({
  transport: custom(provider.provider as any),
});

console.log(await client1.getAddresses(), await client1.getChainId());
```
@frankfka I hope this can help 🙂

**Reply by frankfka** [reply]
I also found that this worked:
```
const publicClient = usePublicClient({ chainId: clientEnv.primaryChain.id });
const { data: walletClient } = useWalletClient();

const primaryProvider = useMemo(() => {
  return new providers.Web3Provider(publicClient as any);
}, [publicClient]);
const signer = useMemo(() => {
  if (!walletClient) {
    return;
  }
  const provider = new providers.Web3Provider(walletClient as any);
  return provider.getSigner(walletClient.account.address);
}, [walletClient]);
```

What does passing in `transport` do as opposed to just passing in `client`?

**Reply by chybisov** [reply]
Oh yeah, that can work as well when we cast to `any`, with `transport` the types are not messed up, but it's pretty much the same setup 🙂

**Reply by frankfka** [reply]
I actually found that ethers v5 didn't like the typing even with `client.transport`. Were you using ethers v6?

**Reply by chybisov** [reply]
@frankfka my example is for ethers v5, but maybe you use different transport than `window.ethereum`?

**Reply by frankfka** [reply]
Interesting, I was just using `usePublicClient()`, but i'll look into this

**Comment by frankfka** [comment]
Are there plans to build in ethers <> viem conversions directly into viem (or in an extension package?) I think a lot of people would find this useful

**Reply by tmm** [reply]
Check out [`ethersWalletToAccount`](https://viem.sh/docs/migration-guide.html#for-ethers-wallet-adapter-use-etherswallettoaccount).

**Reply by kasparkallas** [reply]
Note to future readers that this was already removed.

**Reply by tmm** [reply]
Correct! You can find the reference implementation to copy-paste here: https://viem.sh/docs/migration-guide.html#removed-etherswallettoaccount

**Comment by jxom** [comment, marked as answer]
We now have reference implementations on the wagmi docs!

https://wagmi.sh/react/ethers-adapters

**Comment by paintoshi** [comment]
@jxom Any ideas how to convert into an EIP-1193 compatible provider used by this widget https://github.com/1inch/embedded-widget

I implemented the ethers adapters to get the provider and signer. Then used this Eip1193Bridge from here https://www.npmjs.com/package/@ethersproject/experimental

But I still get incompatibility between them:
![image](https://github.com/wagmi-dev/wagmi/assets/88572775/8ea081fd-8b5a-487d-b423-a973f784975a)

![image](https://github.com/wagmi-dev/wagmi/assets/88572775/4e9ef87e-7223-47ca-8e84-e1115998ddf1)

I also got a typescript error in the wagmi adapters, so maybe I'm doing something wrong.
![image](https://github.com/wagmi-dev/wagmi/assets/88572775/5e37cc86-3d66-4a8b-8cfe-d6810624275b)

**Reply by paintoshi** [reply]
I switched widget to use this instead: https://docs.uniswap.org/sdk/swap-widget/guides/getting-started
It has at least support for wagmi 0.11, but the useSigner() does no longer exist so can't use that. I tried with the eip1193Bridge as mentioned above but they don't allow it as seen here: 
![image](https://github.com/wagmi-dev/wagmi/assets/88572775/b9a35dfd-10a4-4f53-8111-4d74e1ea4f2d)

But I figured it out by using a third variant I call useEthersWeb3Provider, not mentioned in the https://wagmi.sh/react/ethers-adapters

```
export function walletClientToProvider(walletClient: WalletClient) {
  const { chain, transport } = walletClient
  const network = {
    chainId: chain.id,
    name: chain.name,
    ensAddress: chain.contracts?.ensRegistry?.address,
  }
  const provider = new providers.Web3Provider(transport as any, network)
  return provider
}

/** Hook to convert a viem Wallet Client to an ethers.js Web3Provider. */
export function useEthersWeb3Provider({ chainId }: { chainId?: number } = {}) {
  const { data: walletClient } = useWalletClient({ chainId })
  return React.useMemo(
      () => (walletClient ? walletClientToProvider(walletClient) : undefined),
      [walletClient],
  )
}
```

It was exactly what the widget expected.

**Reply by chybisov** [reply]
You can also take a look at this widget https://github.com/lifinance/widget 

We are moving to `wagmi` in the next major soon, but you can also use walletClient -> ethers conversion. 🙂

**Reply by paintoshi** [reply]
@chybisov I'm looking at integrating the lify widget into our game where we use wagmi as wallet controller. How to control the widget using a provider and not the built-in wallet support? I can't find any references of that in the documentation. Edit: After looking some more I can see it's likely the signer param in the WidgetWalletManagement. However my react app crashes hard when trying to use the widget so not sure I'm able to work around that. I'm using react-app-rewired, so maybe that.
